### PR TITLE
Remove even more redundant variables to avoid inconsistent states. Fix another swap bug.

### PR DESCRIPTION
**Pull request by C0D3D3V**
See https://youtu.be/jghPB-Y2BsQ

Please review it carefully. 

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [x] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

File: src/game/client/prediction/entities/character.h [hunk 1]
```diff
@@ -143,6 +130,13 @@ class CCharacter : public CEntity
 	void ResetPrediction();
 	void SetTuneZone(int Zone);
 
+	bool CanNotHammerHit() { return m_Core.m_NoHammerHit; }
+	bool CanNotShotgunHit() { return m_Core.m_NoShotgunHit; }
+	bool CanNotLaserHit() { return m_Core.m_NoLaserHit; }
+	bool CanNotGranadeHit() { return m_Core.m_NoGrenadeHit; }
```
[the remark this answers is not in the document]
Comment (C0D3D3V) [reply src/game/client/prediction/entities/character.h]: my spelling mistakes, terrible. Thank you.

**Comment by C0D3D3V** [comment]
The swap bug was reported on discord by @cyberfig :
If you swap with a partner that is live frozen, the state gets inconsistent, the live frozen player is normal displayed, the normal player is live frozen displayed.

I have not checked if you can exploit the other incontinent variables for something.

**Comment by C0D3D3V** [comment]
>  I'd prefer the affirmative version (CanXXXHit).

And what about the variables themself? 
```
m_NoHammerHit 
m_NoShotgunHit
m_NoGrenadeHit
m_NoLaserHit
```

I left it like this because I also chose `m_NoCollision` and `m_NoHookHit` for the last inconsistency bug because they sound more detailed/verbose to me. Also the default for `m_NoHammerHit`, ... is 0 so, it feels more natural.

**Comment by Chairn** [comment]
Is it possible to fix the swap bug independently of the redundant variable removal?

**Comment by C0D3D3V** [comment]
Sure. But why do you want to keep redundant variables? You just have to keep the redundant variables consistent.

**Comment by C0D3D3V** [comment]
swapping if you are super does not work anyways. Even if both players are super you get this message: 
grafik

**Comment by C0D3D3V** [comment]
I have successfully swapped all ddrace effects. So if I have not overlooked anything else this PR can be merged.

One thing I noticed when testing, but it also happens without the PR. For some strange reason it can happen that two players are denied swapping with the message "You and other player need to have started the map" although both have already started and have already swapped. I can't reproduce this but it happened to me twice. Once with the PR and once without.

**Comment by C0D3D3V** [comment]
@Chairn Do you still want this change? Or is this mergeable?

**Comment by Chairn** [comment]
Its mergeable.

File: src/game/server/entities/character.h [hunk 2]
```diff
@@ -252,6 +238,13 @@ class CCharacter : public CEntity
 	bool HasTelegunGrenade() { return m_Core.m_HasTelegunGrenade; }
 	bool HasTelegunLaser() { return m_Core.m_HasTelegunLaser; }
 
+	bool CanNotHammerHit() { return m_Core.m_NoHammerHit; }
+	bool CanNotShotgunHit() { return m_Core.m_NoShotgunHit; }
+	bool CanNotLaserHit() { return m_Core.m_NoLaserHit; }
+	bool CanNotGrenadeHit() { return m_Core.m_NoGrenadeHit; }
```
Comment (heinrich5991) [inline src/game/server/entities/character.h:244]: CanNot → Cannot, because cannot is a single word IIRC.

Comment (Chairn) [reply src/game/server/entities/character.h:244]: I'd even remove "cannot" and just keep "noXxxHit"

**Comment by edg-l** [comment]
I dislike a lot the `!pOwnerChar->CanNotLaserHit()` you need like 30 secs to process what this means.

 I think `LaserHitDisabled()` sounds better

**Comment by C0D3D3V** [comment]
xD so we have now 3 options:

1. `LaserHitDisabled()`
2. `NoLaserHit()`
3. `CannotLaserHit()`

I think option 1 is the best (Although the variable itself is called `m_NoLaserHit`). What do you think?

**Comment by edg-l** [comment]
I think its fine to use the `!` with the option 1. You can clearly read it as "not disabled"

**Comment by C0D3D3V** [comment]
Ready for final review

**Comment by Learath2** [comment]
Does anyone know the origin of the redundancy? Why were things put into both the core and the character itself? Are we accidentally unfixing a subtle spec or pause bug somewhere?

**Comment by C0D3D3V** [comment]
Except of the live freeze bug that gets fixed with this. The only redundant DDNet variables that were also used inside of gamecore are m_CollisionDisabled and m_HookHitDisabled, and these two redundancys got fixed by me already in an other PR. The other redundant variables were just wrong but unused.

I did not look into the history yet of the redundant variables. But I guess original it was tried to seperate new DDNet variables from vanilla gamecore. And then someone for some reason started to add these variables back to gamecore. 

I personally think it would also be ok to only have these variables in character.cpp but since our characer is split into client/server that would not be so nice (because we have always to define them twice). This way with the variables only in gamecore, we have the same physics variables in client and server and only the extra variables in character.cpp for extra ddnet stuff that is not directly shared with the client. 
An even better solution would be to put all ddnet related physics variables in an DdnetGameModel. So that other mods could instead use their ModGameModel. 

But for now I guess to remove the redundancy is a good start.

**Comment by C0D3D3V** [comment]
rebased

**Pull request merged**